src/Time.mo

```diff
@@ -0,0 +1,13 @@
+/// System time
+
+import Prim "mo:prim";
+module {
+    /// Current system time given as nanoseconds since 1970-01-01. The system guarantees that
```

Comment:
This should also define a type alias for Time, so users don't need Prim.

Reply:
There is no type here to import (do you mean `Word64`), so users don't need `Prim`.

Reply:
Ah, right. So let's define `type Time = Nat64`.

Reply:
Shouldn't we got for `Time = Nat`, and convert in `Time.now()`? The `Nat64` seems to be a rather low-level artifact of the system api.

Reply:
Hm, good question. I wonder how useful that is. Most APIs consuming time values will probably be limited to Nat64, so what would one practically do with larger values?

Reply:
Talk about things past the year 2554 for example :-)

Actually, we probably want `Int`, else we cannot express time before 1970…

The main point is that a high-level language like Motoko should use `Nat` or `Int` everywhere, as the other types are tedious to use (you’d have to put type annotations around literals in arithmetic in some places) and you can run into traps.

Reply:
I suppose you're right, I just wonder what to do with out-of-range times at the boundary to the system or other canisters. Or in other words, is the time type a platform abstraction or a Motoko abstraction?

Reply:
So far, we only pass time from system to motoko, not back, so we don't have to worry about out-of-range (unless the internet computer lives past 2554). But now I wonderif this is short-sighted and we will have to pass time to the system as well… but likely only in things like “timeout”.

Note that on the HTTP Interface we use CBOR encoding, which doesn't distinguish between nat64 and nat, so we are fine there.

I would treat the nat64 here as an Wasm-induced cludge (beause there is no primitive wasm bignum ).

Reply:
Using `Nat` or `Int` would give users a wrong impression that things will be fine after 2554. I think what we really need is an opaque type `Time.t` that hides the implementation detail. While we cannot hide types for now, giving a type alias and keep the signature as `nat64` make more sense to me.